### PR TITLE
java-microprofile: Use LMP 3.1, Use LMP dev mode, use liberty 19.0.0.9

### PR DESCRIPTION
**Pull request by uberskigeek**
### Checklist:

- [x] Read the [Code of Conduct](https://github.com/appsody/website/blob/master/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md).

- [x] Followed the [commit message guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md#commit-message-guidelines).

- [x] Stack adheres to [Appsody stack structure](https://github.com/appsody/website/blob/master/content/docs/stacks/stack-structure.md).

### Modifying an existing stack:

- [x] Updated the stack version in `stack.yaml`

### Contributing a new stack:

- Describe how application dependencies are managed:

- Explain how Appsody file watcher is utilized:

- Describe other Appsody environment variables defined by the stack image:

- Describe any limitations and known issues:

### CODE REVIEW

File: incubator/java-microprofile/templates/default/pom.xml [hunk 1]
```diff
@@ -8,7 +8,7 @@
     <parent><!--required parent POM-->
         <groupId>dev.appsody</groupId>
         <artifactId>java-microprofile</artifactId>
-        <version>[0.2, 0.3)</version>
+        <version>[0.2, 0.4)</version>
```
Comment (scottkurz) [inline incubator/java-microprofile/templates/default/pom.xml]: Should be: [0.3, 0.4) right ?

File: incubator/java-microprofile/templates/default/pom.xml [hunk 2]
```diff
@@ -102,7 +102,6 @@
                     <serverName>defaultServer</serverName>
```
Comment (scottkurz) [inline incubator/java-microprofile/templates/default/pom.xml]: Maybe there's a case to be explicit sometimes even when using default values for these config parms... but I'd particularly question the **packageFile** and **include** parms... are we just copying these over or did we review them?

File: incubator/java-microprofile/image/Dockerfile-stack [hunk 3]
```diff
@@ -30,12 +30,10 @@ ENV APPSODY_WATCH_REGEX="^.*(.xml|.java|.properties)$"
 
 ENV APPSODY_INSTALL="../validate.sh && mvn -B -Dmaven.repo.local=/mvn/repository install -DskipTests && /project/install-dev-deps.sh"
 
-ENV APPSODY_RUN="JVM_ARGS='-javaagent:resources/javametrics-agent.jar' mvn -B -Dmaven.repo.local=/mvn/repository liberty:run"
-ENV APPSODY_RUN_ON_CHANGE="JVM_ARGS='-javaagent:resources/javametrics-agent.jar' mvn -Dmaven.repo.local=/mvn/repository package -DskipTests"
+ENV APPSODY_RUN="JVM_ARGS='-javaagent:resources/javametrics-agent.jar' mvn -B -Dmaven.repo.local=/mvn/repository liberty:dev -DskipTests"
```
Comment (BarDweller) [inline incubator/java-microprofile/image/Dockerfile-stack]: originally, the mp stack ran validate before each operation, this was flipped at somepoint to only run validate at init.
 
this means the parent pom is only being enforced when the project is cut from the template, which is kinda pointless, as you'd really think you could check that as part of the stack build, rather than at init time.. 

As-is, the end user of the project can just nuke the parent pom reference, change all the deps, liberty level etc, and nothing will prevent it..

File: incubator/java-microprofile/image/project/Dockerfile [hunk 4]
```diff
@@ -1,23 +1,34 @@
...
+RUN cd /project && mvn -B install dependency:go-offline -DskipTests
 
+# Copy and run a simple version check
+COPY ./util /project/util
 RUN  /project/util/check_version build
 
-RUN cd /project && mvn -B install dependency:go-offline -DskipTests
+# Copy the user's application pom and resolve all dependencies
+COPY ./user-app/pom.xml /project/user-app/pom.xml
+RUN cd /project/user-app && mvn package -Dskip=true -DskipTests
```
Comment (BarDweller) [inline incubator/java-microprofile/image/project/Dockerfile]: Similar to the dockerfile-stack, validate.sh should be being driven here, to ensure the stack only builds final images that are based on maven projects that still have the parent reference intact.

File: incubator/java-microprofile/image/project/Dockerfile [hunk 5]
```diff
@@ -1,23 +1,34 @@
...
     apt-get install -y maven unzip wget xmlstarlet
 
-COPY . /project
+# Copy and build the dev.appsody:java-microprofile parent pom
+COPY ./pom.xml /project/pom.xml
+RUN cd /project && mvn -B install dependency:go-offline -DskipTests
 
+# Copy and run a simple version check
+COPY ./util /project/util
 RUN  /project/util/check_version build
```
Comment (BarDweller) [inline incubator/java-microprofile/image/project/Dockerfile]: This isn't running the check, it's just building the check_version utility, that's invoked via the validate.sh, except, validate.sh has fallen off all paths except init, where it's a tad pointless.

File: incubator/java-microprofile/image/Dockerfile-stack [hunk 6]
```diff
@@ -30,12 +30,10 @@ ENV APPSODY_WATCH_REGEX="^.*(.xml|.java|.properties)$"
```
Comment (BarDweller) [inline incubator/java-microprofile/image/Dockerfile-stack]: above.. is mounting the m2 repo, you should really also be adding
ENV APPSODY_USER_RUN_AS_LOCAL=true
otherwise any manipulation of the m2 repo during use of the stack during run/debug/test will leave root owned files in the users m2 repo, which is a BAD idea.

File: incubator/java-microprofile/image/Dockerfile-stack [hunk 7]
```diff
@@ -30,12 +30,10 @@ ENV APPSODY_WATCH_REGEX="^.*(.xml|.java|.properties)$"
...
 ENV APPSODY_INSTALL="../validate.sh && mvn -B -Dmaven.repo.local=/mvn/repository install -DskipTests && /project/install-dev-deps.sh"
 
-ENV APPSODY_RUN="JVM_ARGS='-javaagent:resources/javametrics-agent.jar' mvn -B -Dmaven.repo.local=/mvn/repository liberty:run"
-ENV APPSODY_RUN_ON_CHANGE="JVM_ARGS='-javaagent:resources/javametrics-agent.jar' mvn -Dmaven.repo.local=/mvn/repository package -DskipTests"
+ENV APPSODY_RUN="JVM_ARGS='-javaagent:resources/javametrics-agent.jar' mvn -B -Dmaven.repo.local=/mvn/repository liberty:dev -DskipTests"
 ENV APPSODY_RUN_KILL=false
 
-ENV APPSODY_DEBUG="JVM_ARGS='-javaagent:resources/javametrics-agent.jar' mvn -B -Dmaven.repo.local=/mvn/repository liberty:debug"
-ENV APPSODY_DEBUG_ON_CHANGE="JVM_ARGS='-javaagent:resources/javametrics-agent.jar' mvn -Dmaven.repo.local=/mvn/repository package -DskipTests"
+ENV APPSODY_DEBUG="JVM_ARGS='-javaagent:resources/javametrics-agent.jar' mvn -B -Dmaven.repo.local=/mvn/repository liberty:devi -DskipTests"
```
Comment (ericglau) [inline incubator/java-microprofile/image/Dockerfile-stack]: - Typo: `liberty:devi` should be `liberty:dev`

File: incubator/java-microprofile/templates/default/pom.xml (hunk 2 above)
Comment (uberskigeek) [reply incubator/java-microprofile/templates/default/pom.xml]: Just copied them over. Did not review them

File: incubator/java-microprofile/image/Dockerfile-stack (hunk 3 above)
Comment (uberskigeek) [reply incubator/java-microprofile/image/Dockerfile-stack]: Good point. Will have to look at this. The developer experience whenever a code change is made is very slow if package is done every time. Removing the run on change allows just the app restart. Very nice and very quick. Will look into the run on change and see if it can be more selective as to when it does a package.

Comment (scottkurz) [reply incubator/java-microprofile/image/Dockerfile-stack]: Hadn't appreciated the validate routine and how it fit in.

Maybe we could prototype watching just the pom.xml and validating on a pom.xml change.. separately and off-to-the-side of LMP dev mode reacting to the POM change.  It could get a little confusing if you have two "processes" at once reacting to a single change, so let's try and see what that would look like.

File: incubator/java-microprofile/image/project/Dockerfile [hunk 8]
```diff
@@ -14,7 +14,8 @@ RUN  /project/util/check_version build
 
 # Copy the user's application pom and resolve all dependencies
 COPY ./user-app/pom.xml /project/user-app/pom.xml
-RUN cd /project/user-app && mvn package -Dskip=true -DskipTests
+COPY ./validate.sh /project/user-app/validate.sh
+RUN cd /project/user-app && ./validate.sh  && mvn package -Dskip=true -DskipTests
```
Comment (scottkurz) [inline incubator/java-microprofile/image/project/Dockerfile]: I realize the `-Dskip=true` came from #485  not from your PR... but does that really do anything?   Was it supposed to be something else...`-DskipLibertyPackage` maybe?

**Comment by scottkurz** [comment]
A couple more comments:
* Why  not the latest OL version: 19.0.0.11
* From #479, should we change to `mvn -B` in **Dockerfile**?

File: incubator/java-microprofile/image/Dockerfile-stack (hunk 6 above)
Comment (uberskigeek) [reply incubator/java-microprofile/image/Dockerfile-stack]: I added this. The only issue is now the pom.xml and src are owned by the UID and the target directory is owned by root. Prevents the classes directory from being created.. trying to work around this..

File: incubator/java-microprofile/image/Dockerfile-stack [hunk 9]
```diff
@@ -21,21 +21,20 @@ RUN mvn -Dmaven.repo.local=/mvn/repository install dependency:go-offline -DskipT
...
 ENV APPSODY_INSTALL="../validate.sh && mvn -B -Dmaven.repo.local=/mvn/repository install -DskipTests && /project/install-dev-deps.sh"
 
-ENV APPSODY_RUN="JVM_ARGS='-javaagent:resources/javametrics-agent.jar' mvn -B -Dmaven.repo.local=/mvn/repository liberty:run"
-ENV APPSODY_RUN_ON_CHANGE="JVM_ARGS='-javaagent:resources/javametrics-agent.jar' mvn -Dmaven.repo.local=/mvn/repository package -DskipTests"
+ENV APPSODY_RUN="JVM_ARGS='-javaagent:resources/javametrics-agent.jar' mvn -B -Dmaven.repo.local=/mvn/repository liberty:dev -DskipTests"
 ENV APPSODY_RUN_KILL=false
 
-ENV APPSODY_DEBUG="JVM_ARGS='-javaagent:resources/javametrics-agent.jar' mvn -B -Dmaven.repo.local=/mvn/repository liberty:debug"
-ENV APPSODY_DEBUG_ON_CHANGE="JVM_ARGS='-javaagent:resources/javametrics-agent.jar' mvn -Dmaven.repo.local=/mvn/repository package -DskipTests"
+ENV APPSODY_DEBUG="JVM_ARGS='-javaagent:resources/javametrics-agent.jar' mvn -B -Dmaven.repo.local=/mvn/repository liberty:dev -DskipTests"
```
Comment (ericglau) [inline incubator/java-microprofile/image/Dockerfile-stack]: Once https://github.com/appsody/appsody/issues/670 is fixed, we should remove `-DskipTests` from `APPSODY_RUN` and `APPSODY_DEBUG` so that when using appsody with `--interactive` mode, dev mode will let users press Enter to run tests.  

We should wait for that issue to be fixed first before making this change though.  If needed, this can be a separate PR for this stack.

Comment (ericglau) [reply incubator/java-microprofile/image/Dockerfile-stack]: https://github.com/appsody/appsody/issues/670 is fixed in Appsody 0.5.1.  Please proceed to remove `-DskipTests` as mentioned above.  Then if using non-interactive mode, tests will run upon code changes.  If using interactive mode, tests will run on demand when pressing Enter.

**Comment by Kamran64** [comment]
Hey guys, is this ready to be taken off draft or is it still a work in progress?

**Comment by uberskigeek** [comment]
This is no longer valid these changes are in PR 552 "Experimental java openliberty"
